Refactor cli (#702)
fermyon/developer: content/spin/cli-reference.md

url = "https://github.com/fermyon/developer/blob/main/content/spin/cli-reference.md"

---
- [spin add](#spin-add)
- [spin build](#spin-build)
- [spin cloud](#spin-cloud)
- [spin deploy](#spin-deploy)
- [spin cloud deploy](#spin-cloud-deploy)

> - [spin login](#spin-login)

Reviewer: I just noticed `deploy` and `login` are out of alphabetical order.  For findability, they should remain in order.

url = "https://github.com/fermyon/developer/blob/main/content/spin/cli-reference.md"

---
- [spin add](#spin-add)
- [spin build](#spin-build)
- [spin cloud](#spin-cloud)

> - [spin deploy](#spin-deploy)

Reviewer: You need to revise the alphabet song I fear - `deploy` should come after _all_ cloud commands.

{{ blockEnd }}

> {{ startTab "v1.2.0/1.2.1"}}

Author: Yeah - 1.2 is the only version we've released with a minor patch, and it felt wrong to only have 1.2.0 in there. So I agree using just major.minor, until... :-)

{{ blockEnd }}

{{ startTab "v1.3"}}

The `spin cloud` command is implemented by the [Fermyon Cloud Plugin](../cloud/cloud-command-reference).

{{ blockEnd }}

{{ blockEnd }}

<!-- markdownlint-disable-next-line titlecase-rule -->
## spin deploy

> `spin deploy` is a shotcurt to [`spin cloud deploy`](#cloud-deploy).

Reviewer: ```suggestion
`spin deploy` is a shortcut to [`spin cloud deploy`](#cloud-deploy).
```

Reviewer: The link might be wrong too?  
```suggestion
`spin deploy` is a shortcut to [`spin cloud deploy`](#spin-cloud-deploy).
```

Though that just links on to the cloud command reference so maybe better to link directly from here to `/cloud/cloud-command-reference#spin-cloud-deploy`

Author: That would only be true for 1.3, so leaving this for older versions.

{{ blockEnd }}

{{ startTab "v1.3.0"}}

The `spin cloud deploy` command is implemented by the [Fermyon Cloud Plugin](../cloud/cloud-plugin-reference#spin-cloud-deploy).

{{ blockEnd }}

{{ blockEnd }}

<!-- markdownlint-disable-next-line titlecase-rule -->
## spin login

> Please note: the previous `spin login` command (from versions before Spin v0.9.0) has been kept to ensure backward compatibility. In the Spin v0.9.0 release, both the `spin login --help` and [`spin cloud login --help`](#cloud-login) commands will produce the same output.

Author: Good point, I'll remove this.

{{ blockEnd }}

{{ startTab "v1.3.0"}}

The `spin cloud login` command is implemented by the [Fermyon Cloud Plugin](../cloud/cloud-plugin-reference#spin-cloud-login).

{{ blockEnd }}

{{ blockEnd }}

<!-- markdownlint-disable-next-line titlecase-rule -->

> ## spin cloud variables

Author: I hear you, and I think the whole "move from spin to plugin" is not doing us good here. I find it hard to strike the perfect balance, so I think it's better to over-communicate here, than keep consistent.

Once we get to 2.0, we could land on just documenting the top-level plugin command here - e.g., `spin cloud` and refer to the plugin doc. But until then, it's feels weird to have the plugin being half-documented in this reference. My fear is that some will not find the command in this reference, and then be lost.

Potentially have a single section for all the aliases and shortcuts we have, one we pull theses out.

> ## spin cloud variables

Reviewer: Your comment about over-communication is well taken - let's go for it.  We always have the option to iterate.

```

{{ blockEnd }}

{{ startTab "v1.1.0"}}

<!-- @selectiveCpy -->

```console
$ spin help      

spin 1.1.0 
The Spin CLI

USAGE:
    spin <SUBCOMMAND>

OPTIONS:
    -h, --help       Print help information
    -V, --version    Print version information

SUBCOMMANDS:
    add          Scaffold a new component into an existing application
    build        Build the Spin application
    cloud        Commands for publishing applications to the Fermyon Platform
    deploy       Package and upload an application to the Fermyon Platform
    help         Print this message or the help of the given subcommand(s)
    login        Log into the Fermyon Platform
    new          Scaffold a new application based on a template
    plugins      Install/uninstall Spin plugins
    registry     Commands for working with OCI registries to distribute applications
    templates    Commands for working with WebAssembly component templates
    up           Start the Spin application
    watch        Rebuild and restart the Spin application when files changes
```

{{ blockEnd }}

{{ startTab "v1.2.0/1.2.1"}}

<!-- @selectiveCpy -->

```console
$ spin help      

> spin 1.2.0/1.2.1 

Author: Good catch!

    build        Build the Spin application
    cloud        Commands for publishing applications to the Fermyon Cloud.
    deploy       Package and upload an application to the Fermyon Cloud.
    help         Print this message or the help of the given subcommand(s)
    login        Log into the Fermyon Cloud.
    new          Scaffold a new application based on a template
    plugins      Install/uninstall Spin plugins
    registry     Commands for working with OCI registries to distribute applications
    templates    Commands for working with WebAssembly component templates
    up           Start the Spin application
    watch        Build and run the Spin application, rebuilding and restarting it when files
                     change

```

{{ blockEnd }}

{{ blockEnd }}

> Please note: Spin `help` is a convenient way to access help using a subcommand, instead of using the `--help` option. For example, `spin help cloud` will give you the same output as `spin cloud --help`. Similarly, `spin help build` will give you the same output as `spin build --help` and so forth.

<!-- markdownlint-disable-next-line titlecase-rule -->
## spin login

> `spin login` is a shotcurt to [`spin cloud login`](#cloud-login).

Reviewer: ```suggestion
`spin login` is a shortcut to [`spin cloud login`](/cloud/cloud-command-reference#spin-cloud-login).
```

    registry     Commands for working with OCI registries to distribute applications
    templates    Commands for working with WebAssembly component templates
    up           Start the Spin application
    watch        Build and run the Spin application, rebuilding and restarting it when files
                     change
```

{{ blockEnd }}

{{ startTab "v1.3.0"}}

<!-- @selectiveCpy -->

```console
$ spin help

> spin 1.3.0 (9fb8256 2023-06-12)

Author: Looks like we've always done this, just not having it in this reference.

For ease of updating going forward, I'll leave this in, as there will be one less special rule to remember for the next release :-)

**Note:** For additional information, please see the [Managing Plugins](/spin/managing-plugins) and/or [Creating Plugins](/spin/plugin-authoring) sections of the documentation.

<!-- markdownlint-disable-next-line titlecase-rule -->

> ## spin oci

Author: Oh - I see it's hidden.

Will remove this section.

{{ blockEnd }}

> {{ startTab "v1.2"}}

Reviewer: I think `spin registry push` might have lost its 1.3 tab?

{{ blockEnd }}

<!-- markdownlint-disable-next-line titlecase-rule -->

> ## spin template

Author: You really don't want those aliases in there :-)

Removed...

{{ blockEnd }}

{{ startTab "v1.1"}}

<!-- @selectiveCpy -->

```console

> $ spin templates upgrade --help1

Reviewer: ```suggestion
$ spin templates upgrade --help
```

The files of the templates are copied to the local template store: a directory in your data or home
directory.

USAGE:
    spin templates upgrade [OPTIONS]

OPTIONS:
        --all
            By default, Spin displays the list of installed repositories and prompts you to choose
            which to upgrade.  Pass this flag to upgrade all repositories without prompting

        --branch <BRANCH>
            The optional branch of the git repository, if a specific repository is given

    -h, --help
            Print help information

        --repo <GIT_URL>
            By default, Spin displays the list of installed repositories and prompts you to choose
            which to upgrade.  Pass this flag to upgrade only the specified repository without
            prompting
```

{{ blockEnd }}

{{ blockEnd }}

**Note:** For additional information, please see the [Managing Templates](/spin/managing-templates) and/or [Creating Templates](/spin/template-authoring) sections of the documentation.

<!-- markdownlint-disable-next-line titlecase-rule -->
## spin up

The following options are available in relation to running your Spin application.

> Note: There are three trigger options which only applies to the HTTP trigger (`--listen`, `--tls-cert` and `--tls-key`).

Author: I can make a `Spin up (HTTP Trigger)` heading with the extended options. What do you think about that?
<img width="1186" alt="image" src="https://github.com/fermyon/developer/assets/12242001/fa3c9ec3-03ac-4305-95ff-31be9b68da4c">

<img width="864" alt="image" src="https://github.com/fermyon/developer/assets/12242001/68de6e15-5d54-4aeb-9b11-82f38d06f7bc">

<img width="311" alt="image" src="https://github.com/fermyon/developer/assets/12242001/2d915cf6-5156-4e93-aad3-44b4b4649200">

> Note: There are three trigger options which only applies to the HTTP trigger (`--listen`, `--tls-cert` and `--tls-key`).

Reviewer: That's pretty much what we do at the moment.  The current one shows the _full_ list of flags under each trigger: is that what you propose, or only to show the additional flags for the trigger?  (I would prefer "additional flags only," but it depends how close you want to stick to "here is a printout of the help.")  Anyway yes that sounds fine.

(I can't tell if that's a subheading or a top-level heading - my suggestion would be subheading under `spin up`.)

> Note: There are three trigger options which only applies to the HTTP trigger (`--listen`, `--tls-cert` and `--tls-key`).

Author: 👍🏻 - I did make it a subheading, but the ToC doesn't render it that way. Will open an issue on it.

    -L, --log-dir <APP_LOG_DIR>
            Log directory for the stdout and stderr of components

>         --listen <ADDRESS>

Reviewer: I thought the HTTP options were being shown in a separate subheading?

Author: That was a miss, fixed now.

            used. The cert should be in PEM format
            
            [env: SPIN_TLS_CERT=]
            Only appplies to HTTP triggers

        --tls-key <TLS_KEY>
            The path to the certificate key to use for https, if this is not set, normal http will
            be used. The key should be in PKCS#8 format
            
            [env: SPIN_TLS_KEY=]
            Only appplies to HTTP triggers
```

{{ blockEnd }}

{{ blockEnd }}

<!-- markdownlint-disable-next-line titlecase-rule -->
### spin up (HTTP Trigger)

> The following additional trigger options are available for the [spin up](#spin-up) command, when using the HTTP trigger. E.g., `spin up --listen`

Reviewer: ```suggestion
The following additional trigger options are available for the [spin up](#spin-up) command, when using the HTTP trigger. E.g. `spin up --listen`
```

```

{{ blockEnd }}

{{ startTab "v1.2"}}

<!-- @selectiveCpy -->

```console
    --listen <ADDRESS>
        IP address and port to listen on

        [default: 127.0.0.1:3000]

    --tls-cert <TLS_CERT>
        The path to the certificate to use for https, if this is not set, normal http will be
        used. The cert should be in PEM format

        [env: SPIN_TLS_CERT=]

    --tls-key <TLS_KEY>
        The path to the certificate key to use for https, if this is not set, normal http will
        be used. The key should be in PKCS#8 format

        [env: SPIN_TLS_KEY=]
```

{{ blockEnd }}

{{ startTab "v1.3"}}

<!-- @selectiveCpy -->

```console
    --listen <ADDRESS>
        IP address and port to listen on

        [default: 127.0.0.1:3000]

    --tls-cert <TLS_CERT>
        The path to the certificate to use for https, if this is not set, normal http will be
        used. The cert should be in PEM format

        [env: SPIN_TLS_CERT=]

    --tls-key <TLS_KEY>
        The path to the certificate key to use for https, if this is not set, normal http will
        be used. The key should be in PKCS#8 format

        [env: SPIN_TLS_KEY=]
```

{{ blockEnd }}

{{ blockEnd }}

> > **Please note:** If the `-f` or `--from` options do not accurately infer the intended registry or `.toml` file for your application, then you can explicitly specify either the `--from-registry` or  `--from-file` options to clarify this.

Reviewer: This belongs in the main section of `spin up` not the HTTP triggers section.

            used. The cert should be in PEM format
            
            [env: SPIN_TLS_CERT=]
            Only appplies to HTTP triggers

        --tls-key <TLS_KEY>
            The path to the certificate key to use for https, if this is not set, normal http will
            be used. The key should be in PKCS#8 format
            
            [env: SPIN_TLS_KEY=]
            Only appplies to HTTP triggers

```

{{ blockEnd }}

{{ blockEnd }}

> > **Please note:** If the `-f` or `--from` options do not accurately infer the intended registry or `.toml` file for your application, then you can explicitly specify either the `--from-registry` or  `--from-file` options to clarify this.

Reviewer: It's explained in an error message if the inference doesn't work.  I'm not hugely enthusiastic about making the help text even longer.

The review comment, though, was in reference to if we were going to manually edit help text body anyway (for the HTTP stuff).  Since that's no longer the plan, I think this comment goes away and we should leave this as is.

{{ blockEnd }}

{{ startTab "v1.3"}}

<!-- @selectiveCpy -->

```console
$ spin watch --help

spin-watch 
Build and run the Spin application, rebuilding and restarting it when files change

USAGE:
    spin watch [OPTIONS] [UP_ARGS]...

ARGS:
    <UP_ARGS>...    Arguments to be passed through to spin up

OPTIONS:
    -c, --clear                       Clear the screen before each run
    -d, --debounce <DEBOUNCE>         Set the timeout between detected change and re-execution, in
                                      milliseconds [default: 100]
    -f, --from <APP_MANIFEST_FILE>    The application to watch. This may be a manifest (spin.toml)
                                      file, or a directory containing a spin.toml file. If omitted,
                                      it defaults to "spin.toml" [default: spin.toml]
    -h, --help                        Print help information
        --skip-build                  Only run the Spin application, restarting it when build
                                      artifacts change
```

{{ blockEnd }}

> {{ blockEnd }}

Reviewer: ~~I might be misreading this but it looks like you got rid of the "which files `spin watch` watches" section.  Was that intentional?~~ You moved it to the discursive section which is probably a better home - thanks!

{{ blockEnd }}

{{ startTab "v1.1"}}

| Command                                                                               | Stability    |
| ------------------------------------------------------------------------------------- | ------------ |
| <code>spin add</code>                                                                 | Stable       |
| <code>spin build</code>                                                               | Stable       |
| <code>spin new</code>                                                                 | Stable       |
| <code>spin plugins <install&vert;list&vert;uninstall&vert;update&vert;upgrade></code> | Stable       |
| <code>spin templates <install&vert;list&vert;uninstall&vert;upgrade></code>           | Stable       |
| <code>spin up</code>                                                                  | Stable       |
| <code>spin cloud <deploy&vert;login></code>                                           | Stabilizing  |
| <code>spin registry</code>                                                            | Stabilizing  |

> | <code>spin watch</code>                                                               | Experimental |

Reviewer: This was given as `stabilising` before - was that an error?  It can't have retrospectively gone backwards can it?!

Reviewer: (Which raises the question of what does it take to count it as "stabilising" - it's given as "experimental" in 1.3.)

Author: It was stabilizing in 1.1, and moved to experimental in 1.2, so I'd assume the 1.1 was an error to begin with. Marking this experimental in this PR. We can fix in a subsequent if this is wrong.
